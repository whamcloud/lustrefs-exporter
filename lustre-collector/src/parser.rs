// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::{
    ldlm, llite, mdd_parser,
    mds::{self, client_count_parser},
    mgs::mgs_parser,
    osd_parser, oss, quota, top_level_parser,
    types::Record,
};
use combine::{choice, error::ParseError, many, Parser, Stream};

pub fn params() -> Vec<String> {
    top_level_parser::top_level_params()
        .into_iter()
        .chain(client_count_parser::params())
        .chain(osd_parser::params())
        .chain(mgs_parser::params())
        .chain(oss::params())
        .chain(mds::params())
        .chain(ldlm::params())
        .chain(llite::params())
        .chain(mdd_parser::params())
        .chain(quota::params())
        .collect()
}

pub fn parse<I>() -> impl Parser<I, Output = Vec<Record>>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many(choice((
        top_level_parser::parse().map(|x| vec![x]),
        client_count_parser::parse(),
        osd_parser::parse().map(|x| vec![x]),
        mgs_parser::parse().map(|x| vec![x]),
        oss::parse().map(|x| vec![x]),
        mds::parse().map(|x| vec![x]),
        ldlm::parse().map(|x| vec![x]),
        llite::parse().map(|x| vec![x]),
        mdd_parser::parse().map(|x| vec![x]),
        quota::parse().map(|x| vec![x]),
    )))
    .map(|xs: Vec<_>| xs.into_iter().flatten().collect())
}

#[cfg(test)]
mod tests {
    use super::*;
    use combine::parser::EasyParser;
    use include_dir::{include_dir, Dir};
    use insta::assert_debug_snapshot;

    static VALID_FIXTURES: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/src/fixtures/valid/");

    #[test]
    fn test_valid_fixtures() {
        for dir in VALID_FIXTURES.find("*").unwrap() {
            match dir {
                include_dir::DirEntry::Dir(_) => {}
                include_dir::DirEntry::File(file) => {
                    let name = file.path().to_string_lossy();

                    let contents = file.contents_utf8().unwrap();

                    let result = parse()
                        .easy_parse(contents)
                        .map_err(|err| err.map_position(|p| p.translate_position(contents)))
                        .unwrap();

                    assert_debug_snapshot!(format!("valid_fixture_{name}"), result);
                }
            }
        }
    }

    #[test]
    fn test_params() {
        assert_debug_snapshot!(params());
    }

    #[test]
    fn test_mdt_output() {
        let x = r#"memused=343719411
memused_max=344830779
lnet_memused=140274209
health_check=healthy
mdt.fs-MDT0000.exports.0@lo.uuid=fs-MDT0000-lwp-MDT0000_UUID
mdt.fs-MDT0000.exports.10.0.2.15@tcp.uuid=
613beb43-5df2-2ace-4209-be66b4b509df
568acc64-085e-ada1-d493-6318930dfa74
mdt.fs-MDT0000.exports.10.73.20.21@tcp.uuid=
fs-MDT0000-lwp-OST000a_UUID
fs-MDT0000-lwp-OST0004_UUID
fs-MDT0000-lwp-OST0000_UUID
fs-MDT0000-lwp-OST0006_UUID
fs-MDT0000-lwp-OST0002_UUID
fs-MDT0000-lwp-OST0008_UUID
mdt.fs-MDT0000.exports.10.73.20.22@tcp.uuid=
fs-MDT0000-lwp-OST0007_UUID
fs-MDT0000-lwp-OST0003_UUID
fs-MDT0000-lwp-OST0009_UUID
fs-MDT0000-lwp-OST0001_UUID
fs-MDT0000-lwp-OST000b_UUID
fs-MDT0000-lwp-OST0005_UUID
mdt.fs2-MDT0000.exports.0@lo.uuid=fs2-MDT0000-lwp-MDT0000_UUID
mdt.fs2-MDT0000.exports.10.0.2.15@tcp.uuid=
a7b7c685-18c1-eecc-eae2-5880f431cae3
6f2afad1-ff3a-cdd0-721f-dcc123cae427
mdt.fs2-MDT0000.exports.10.73.20.12@tcp.uuid=fs2-MDT0000-lwp-OST0002_UUID
mdt.fs2-MDT0000.exports.10.73.20.21@tcp.uuid=
fs2-MDT0000-lwp-OST0003_UUID
fs2-MDT0000-lwp-OST0000_UUID
mdt.fs2-MDT0000.exports.10.73.20.22@tcp.uuid=fs2-MDT0000-lwp-OST0001_UUID
ldlm.namespaces.mdt-fs-MDT0000_UUID.contended_locks=32
ldlm.namespaces.mdt-fs2-MDT0000_UUID.contended_locks=32
ldlm.namespaces.mdt-fs-MDT0000_UUID.contention_seconds=2
ldlm.namespaces.mdt-fs2-MDT0000_UUID.contention_seconds=2
ldlm.namespaces.mdt-fs-MDT0000_UUID.ctime_age_limit=10
ldlm.namespaces.mdt-fs2-MDT0000_UUID.ctime_age_limit=10
ldlm.namespaces.mdt-fs-MDT0000_UUID.early_lock_cancel=0
ldlm.namespaces.mdt-fs2-MDT0000_UUID.early_lock_cancel=0
ldlm.namespaces.mdt-fs-MDT0000_UUID.lock_count=0
ldlm.namespaces.mdt-fs2-MDT0000_UUID.lock_count=0
ldlm.namespaces.mdt-fs-MDT0000_UUID.lock_timeouts=0
ldlm.namespaces.mdt-fs2-MDT0000_UUID.lock_timeouts=0
ldlm.namespaces.mdt-fs-MDT0000_UUID.lock_unused_count=0
ldlm.namespaces.mdt-fs2-MDT0000_UUID.lock_unused_count=0
ldlm.namespaces.mdt-fs-MDT0000_UUID.lru_max_age=3900000
ldlm.namespaces.mdt-fs2-MDT0000_UUID.lru_max_age=3900000
ldlm.namespaces.mdt-fs-MDT0000_UUID.lru_size=800
ldlm.namespaces.mdt-fs2-MDT0000_UUID.lru_size=800
ldlm.namespaces.mdt-fs-MDT0000_UUID.max_nolock_bytes=0
ldlm.namespaces.mdt-fs2-MDT0000_UUID.max_nolock_bytes=0
ldlm.namespaces.mdt-fs-MDT0000_UUID.max_parallel_ast=1024
ldlm.namespaces.mdt-fs2-MDT0000_UUID.max_parallel_ast=1024
ldlm.namespaces.mdt-fs-MDT0000_UUID.resource_count=0
ldlm.namespaces.mdt-fs2-MDT0000_UUID.resource_count=0
mdt.fs-MDT0000.md_stats=
snapshot_time             1583789082.568118366 secs.nsecs
getattr                   4 samples [reqs]
statfs                    2 samples [reqs]
mdt.fs2-MDT0000.md_stats=
snapshot_time             1583789082.568222478 secs.nsecs
getattr                   2 samples [reqs]
statfs                    2 samples [reqs]
mdt.fs-MDT0000.num_exports=15
mdt.fs2-MDT0000.num_exports=7
osd-ldiskfs.fs-MDT0000.filesfree=2621151
osd-ldiskfs.fs2-MDT0000.filesfree=2621167
osd-ldiskfs.fs-MDT0000.filestotal=2621440
osd-ldiskfs.fs2-MDT0000.filestotal=2621440
osd-ldiskfs.fs-MDT0000.kbytesavail=1531876
osd-ldiskfs.fs2-MDT0000.kbytesavail=1532068
osd-ldiskfs.fs-MDT0000.kbytesfree=1793268
osd-ldiskfs.fs2-MDT0000.kbytesfree=1793460
osd-ldiskfs.fs-MDT0000.kbytestotal=1819968
osd-ldiskfs.fs2-MDT0000.kbytestotal=1819968
"#;

        let result = parse()
            .easy_parse(x)
            .map_err(|err| err.map_position(|p| p.translate_position(x)))
            .unwrap();

        assert_debug_snapshot!(result);
    }
}
