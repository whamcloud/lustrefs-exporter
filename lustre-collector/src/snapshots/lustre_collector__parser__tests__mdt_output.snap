---
source: src/parser.rs
expression: result
---
(
    [
        Host(
            Memused(
                HostStat {
                    param: Param(
                        "memused",
                    ),
                    value: 343719411,
                },
            ),
        ),
        Host(
            MemusedMax(
                HostStat {
                    param: Param(
                        "memused_max",
                    ),
                    value: 344830779,
                },
            ),
        ),
        Host(
            LNetMemUsed(
                HostStat {
                    param: Param(
                        "lnet_memused",
                    ),
                    value: 140274209,
                },
            ),
        ),
        Host(
            HealthCheck(
                HostStat {
                    param: Param(
                        "health_check",
                    ),
                    value: HealthCheckStat {
                        healthy: true,
                        targets: [],
                    },
                },
            ),
        ),
        Target(
            ConnectedClients(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "connected_clients",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            ConnectedClients(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "connected_clients",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            ContendedLocks(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "contended_locks",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 32,
                },
            ),
        ),
        Target(
            ContendedLocks(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "contended_locks",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 32,
                },
            ),
        ),
        Target(
            ContentionSeconds(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "contention_seconds",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            ContentionSeconds(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "contention_seconds",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 2,
                },
            ),
        ),
        Target(
            CtimeAgeLimit(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "ctime_age_limit",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 10,
                },
            ),
        ),
        Target(
            CtimeAgeLimit(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "ctime_age_limit",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 10,
                },
            ),
        ),
        Target(
            EarlyLockCancel(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "early_lock_cancel",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            EarlyLockCancel(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "early_lock_cancel",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockCount(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lock_count",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockCount(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lock_count",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockTimeouts(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lock_timeouts",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockTimeouts(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lock_timeouts",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockUnusedCount(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lock_unused_count",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LockUnusedCount(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lock_unused_count",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            LruMaxAge(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lru_max_age",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 3900000,
                },
            ),
        ),
        Target(
            LruMaxAge(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lru_max_age",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 3900000,
                },
            ),
        ),
        Target(
            LruSize(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lru_size",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 800,
                },
            ),
        ),
        Target(
            LruSize(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "lru_size",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 800,
                },
            ),
        ),
        Target(
            MaxNolockBytes(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "max_nolock_bytes",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            MaxNolockBytes(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "max_nolock_bytes",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            MaxParallelAst(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "max_parallel_ast",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 1024,
                },
            ),
        ),
        Target(
            MaxParallelAst(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "max_parallel_ast",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 1024,
                },
            ),
        ),
        Target(
            ResourceCount(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "resource_count",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            ResourceCount(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "resource_count",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 0,
                },
            ),
        ),
        Target(
            Stats(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "md_stats",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: [
                        Stat {
                            name: "getattr",
                            units: "reqs",
                            samples: 4,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "statfs",
                            units: "reqs",
                            samples: 2,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                    ],
                },
            ),
        ),
        Target(
            Stats(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "md_stats",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: [
                        Stat {
                            name: "getattr",
                            units: "reqs",
                            samples: 2,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                        Stat {
                            name: "statfs",
                            units: "reqs",
                            samples: 2,
                            min: None,
                            max: None,
                            sum: None,
                            sumsquare: None,
                        },
                    ],
                },
            ),
        ),
        Target(
            NumExports(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "num_exports",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 15,
                },
            ),
        ),
        Target(
            NumExports(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "num_exports",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 7,
                },
            ),
        ),
        Target(
            FilesFree(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "filesfree",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 2621151,
                },
            ),
        ),
        Target(
            FilesFree(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "filesfree",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 2621167,
                },
            ),
        ),
        Target(
            FilesTotal(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "filestotal",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 2621440,
                },
            ),
        ),
        Target(
            FilesTotal(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "filestotal",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 2621440,
                },
            ),
        ),
        Target(
            KBytesAvail(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "kbytesavail",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 1531876,
                },
            ),
        ),
        Target(
            KBytesAvail(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "kbytesavail",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 1532068,
                },
            ),
        ),
        Target(
            KBytesFree(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "kbytesfree",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 1793268,
                },
            ),
        ),
        Target(
            KBytesFree(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "kbytesfree",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 1793460,
                },
            ),
        ),
        Target(
            KBytesTotal(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "kbytestotal",
                    ),
                    target: Target(
                        "fs-MDT0000",
                    ),
                    value: 1819968,
                },
            ),
        ),
        Target(
            KBytesTotal(
                TargetStat {
                    kind: Mdt,
                    param: Param(
                        "kbytestotal",
                    ),
                    target: Target(
                        "fs2-MDT0000",
                    ),
                    value: 1819968,
                },
            ),
        ),
    ],
    "",
)
